### PR TITLE
Support bech32 encoding without a checksum

### PR DESCRIPTION
**Pull request by jkczyz**
[BOLT 12 Offers](https://github.com/lightning/bolts/pull/798) uses bech32 encoding without a checksum since QR codes already have a checksum. Add functions `encode_without_checksum` and `decode_without_checksum` to support this use case.

Also, remove overall length check in decode since it is unnecessary.

### CODE REVIEW

File: src/lib.rs
Comment (jkczyz) [reply src/lib.rs]: Made a similar wrapper for without a checksum, though there's a small amount of code duplication that way.

**Comment by tcharding** [comment]
Sure thing. Can you squash the first two commits please because the second one is patching the changes in the first. Not super important but I'd put patch 3 at the front since it is a trivial improvement. Thanks.

**Comment by tcharding** [comment]
Hi @jkczyz, thanks for your contribution. This PR does the job but its not pretty. How urgent is this? Is it something you need to use today or is it just something you did for fun?

The reason its not pretty is that it adds functionality that is obviously just plastered on top. By that I mean the whole `Bech32Writer` design is there to track and add the checksum so then adding another type that mirrors it but does not add a checksum is kinda strange. I'm basically making the same observation @TheBlueMatt did above but on this iteration of the PR. If we want a good solution I think we need to dig a bit deeper.

**Comment by jkczyz** [comment]
> Hi @jkczyz, thanks for your contribution. This PR does the job but its not pretty. How urgent is this? Is it something you need to use today or is it just something you did for fun?
> 

Somewhat urgent as it will hold up our work on Offers (https://github.com/lightningdevkit/rust-lightning/pull/1597). I have a draft that uses this through `rust-bitcoin`'s re-export, so would need a new release for both.

> The reason its not pretty is that it adds functionality that is obviously just plastered on top. By that I mean the whole `Bech32Writer` design is there to track and add the checksum so then adding another type that mirrors it but does not add a checksum is kinda strange. I'm basically making the same observation @TheBlueMatt did above but on this iteration of the PR. If we want a good solution I think we need to dig a bit deeper.

I don't actually need to use `Bech32WithoutChecksumWriter` directly, so I'm happy to inline the code. I mainly included for  consistency and for code reuse, which is no longer the case it seems.

**Comment by apoelstra** [comment]
As a stop-gap you could just drop the last 6 characters/u8s from a normal bech32 encoding.

**Comment by jkczyz** [comment]
> Sure thing. Can you squash the first two commits please because the second one is patching the changes in the first. Not super important but I'd put patch 3 at the front since it is a trivial improvement. Thanks.

All done.

**Comment by jkczyz** [comment]
> As a stop-gap you could just drop the last 6 characters/u8s from a normal bech32 encoding.

For encoding that would work, but we would still fail to decode a bech32 offer since it doesn't have a checksum.

File: src/lib.rs
Comment (jkczyz) [inline src/lib.rs:453]: I'm a little uncertain about this. Seems to maintain the same behavior, but it is odd that an `fmt::Error` is always wrapped with `Ok`.

Comment (apoelstra) [reply src/lib.rs:453]: Yeah, the existing API is really weird.

Comment (apoelstra) [reply src/lib.rs:453]: Why are you constructing a closure and then immediately calling it?

Comment (jkczyz) [reply src/lib.rs:453]: Mainly as a simple way to maintain the API behavior. i.e., wrapping a `Err(fmt::Error)` returned in each of the three places using `?` with an `Ok`.  Without the closure, it would be:

```rust
    if let Err(e) = fmt.write_str(&hrp) { return Ok(Err(e)); }
    if let Err(e) = fmt.write_char(SEP) { return Ok(Err(e)); }
    for b in data.as_ref() {
        if let Err(e) = fmt.write_char(b.to_char()) { return Ok(Err(e)); }
    }
    Ok(Ok(()))
```

Happy to update the error type is your prefer. It probably should be an enum that is either `bech32::Error` or `fmt::Error`. Or maybe adding a variant to `bech32::Error` for `fmt::Error`.

Comment (apoelstra) [reply src/lib.rs:453]: This whole `Result<Result>` thing is weird to me -- I think we should add `fmt::Error` to the main `Error` enum and then flatten this return type.

I wouldn't worry about maintaining the API here. We want to overhaul this API anyway.

Comment (jkczyz) [reply src/lib.rs]: Flatted the `Result`. Let me know if you prefer a different name for the variant or if you want to drop the wrapped `fmt::Error` as it currently doesn't provide any additional information.

**Comment by apoelstra** [comment]
Loosk good to me! Though I guess CI wants you to add a rustfmt commit.

@Kixunil mind taking another look at this? I think now my remaining unease with the API is just my existing unease at the API, and if we want to clean it up we should just go over the whole library.

ACK 3e95975d7c90450c2735a8cac1c6e95277713371

**Review by apoelstra — APPROVED** [review]
ACK 573d7f4756ed99f996e8ed9b3a70477fa12a3443

**Comment by Kixunil** [comment]
Can't right now, give me a day or so.

File: src/lib.rs
Comment (tcharding) [reply src/lib.rs:509]: I convinced myself everything was ok by adding a test (that passes):
```rust

    #[test]
    fn roundtrip_without_checksum() {
        let hrp = "lnbc";
        let data = "Hello World!".as_bytes().to_base32();

        let encoded = encode_without_checksum(hrp, data.clone()).expect("failed to encode");
        let (decoded_hrp, decoded_data) = decode_without_checksum(&encoded).expect("failed to decode");

        assert_eq!(decoded_hrp, hrp);
        assert_eq!(decoded_data, data);
    }
```

Perhaps the `decode_without_checksum` public function should just call straight through to a helper function `_decode` so the `decode` function can use `_decode` as well and then it will read nicely? Something like

```rust

/// Decode a bech32 string into the raw HRP and the data bytes, assuming no checksum.
///
/// Returns the HRP in lowercase and the data with the checksum removed.
fn decode_without_checksum(s: &str) -> Result<(String, Vec<u5>), Error> {
    _decode(s)
}

/// Decode a bech32 string into the raw HRP and the `u5` data.
pub fn _decode(s: &str) -> Result<(String, Vec<u5>), Error> {
...
```

File: src/lib.rs
Comment (tcharding) [inline src/lib.rs:529]: This line is stale.

**Review by tcharding — APPROVED** [review]
Thanks for sticking with this! Functionality looks correct to me. Perhaps add the unit test I used if you like it. I'm not super happy with the function naming but perhaps we can leave tat till another day, like others have said already this crate is about to get an overhaul anyways. If this solves your usecase I'm happy to ack it.

**Comment by tcharding** [comment]
tACK 573d7f4756ed99f996e8ed9b3a70477fa12a3443

**Comment by jkczyz** [comment]
Addressed all feedback. Can squash fixups whenever you're ready.

**Review by apoelstra — APPROVED** [review]
ACK f768f039b0e136b9a90f711301d3dba0f997c770

**Comment by apoelstra** [comment]
Looks good to me. Yes, would prefer squashing all the fixups.

**Review by apoelstra — APPROVED** [review]
ACK 86edca9371a63dfbce9fed27c00b214b936c19e3

File: src/lib.rs
Comment (Kixunil) [inline src/lib.rs:220]: FTR `std` ignores errors instead of panicking in drop and it is considered the right thing to do because panicking in drop is a footgun.

File: src/lib.rs
Comment (Kixunil) [inline src/lib.rs:441]: It is possible to o this without `Cow` which should be more efficient:

```rust
let hrp_owned; // yes, this is allowed in Rust!
let hrp = match check_hrp(hrp)? {
    Case::Upper => {
        hrp_owned = hrp.to_lowercase();
        hrp_owned.as_str()
    },
    Case::Lower | Case::None => hrp,
};
```

File: src/lib.rs
Comment (Kixunil) [inline src/lib.rs:435]: It'd be better to have `Hrp` newtype to make the reurn type `fmt::Result` allowing use in `Display` implementations.

File: src/lib.rs
Comment (Kixunil) [inline src/lib.rs:672]: I don't think this is the right way of using `fmt::Error`. It's just a marker that something failed and you're supposed to retrieve the error fro underlying stream. In case of `std::io` it's done for you and in case of `String` there's no error so unwrapping is the correct thing to do.

This library abuses the `fmt::Write` API to signal display failure but `Display` must not fail for any other reason than formatter failing.

From [the docs](https://doc.rust-lang.org/std/fmt/index.html):

> Formatting implementations should ensure that they propagate errors from the [Formatter](https://doc.rust-lang.org/std/fmt/struct.Formatter.html) (e.g., when calling [write!](https://doc.rust-lang.org/std/macro.write.html)). However, they should never return errors spuriously. That is, a formatting implementation must and may only return an error if the passed-in [Formatter](https://doc.rust-lang.org/std/fmt/struct.Formatter.html) returns an error. This is because, contrary to what the function signature might suggest, string formatting is an infallible operation. This function only returns a result because writing to the underlying stream might fail and it must provide a way to propagate the fact that an error has occurred back up the stack.

**Review by Kixunil — COMMENTED** [review]
Finally had the time to review it. The API has problems but it was the case before, so not a problem with this particular PR. It just needs to be cleaned up before 1.0.

File: src/lib.rs
Comment (apoelstra) [reply src/lib.rs:220]: Lol, oops! I tried to look for double-panics but got a bit confused and somehow missed this completely obvious one.

But I think we should remove this `Drop` business entirely so it's fine for now. I've never heard of a wrapper struct that takes a `&mut` ref and then has finalization logic on `Drop`...I suspect this was an API experiment that never failed enough to be removed.

File: src/lib.rs
Comment (apoelstra) [reply src/lib.rs:441]: That's a neat lifetime-organizing trick!

File: src/lib.rs
Comment (apoelstra) [reply src/lib.rs:672]: Good find! This will definitely inform our API overhaul.

File: src/lib.rs
Comment (Kixunil) [reply src/lib.rs:220]: IDK, it's not too different from `BufWriter` behavior. But I guess compiler warning would be better than drop. I was thinking of abusing `must_use` but then we couldn't implement any methods taking `&mut self` and it'd be fragile anyway.

**Comment by jkczyz** [comment]
> > As a stop-gap you could just drop the last 6 characters/u8s from a normal bech32 encoding.
> 
> For encoding that would work, but we would still fail to decode a bech32 offer since it doesn't have a checksum.

For sake of planning, what timeframe should I expect for seeing this in a release? Do let me know if I can help in review if other PRs need to be merged first. Just want to set appropriate expectations given this will need to get into a `rust-bitcoin` release, too, before we can use it in `rust-lightning`.

**Comment by tcharding** [comment]
We just dropped a `rust-bitcoin` release so the next one could be six months away. This poses an interesting dilemma, we don't want to be doing major releases of `rust-bitcoin` ever week but if a user is waiting on a merged feature in a dependency that we control it seems rude to hold it up too long.  

If we are going to release then #68 and #72 would likely be nice to consider first.

**Comment by apoelstra** [comment]
We can do a much faster release of this than six months. It's still fine if rust-bitcoin depends on an old version -- `bech32` is pretty self-contained so it should be easy to work with multiple versions of it at once.

**Comment by apoelstra** [comment]
But having said that yes, I'd like to do a bunch more work on this crate before releasing .. so it could be a month or more.

**Comment by jkczyz** [comment]
Gotcha. I think had trouble with multiple versions but can't recall the exact error. Will try again once there's a new `bech32` release. Otherwise, I can unblock the current work by holding off on offer parsing. I made a builder framework in the interim, so we can move forward with dependent work using that.

**Comment by Kixunil** [comment]
Could the code be changed to be non-breaking? Minor version bump wouldn't be so painful and we could do API overhaul later.

**Comment by jkczyz** [comment]
> Could the code be changed to be non-breaking? Minor version bump wouldn't be so painful and we could do API overhaul later.

Are you suggesting doing that in `rust-bech32` and do a similar minor version bump in `rust-bitcoin`?

**Comment by jkczyz** [comment]
Ah, so since `rust-bitcoin` specifies `version = "0.9.0"` for `bech32`, any version `>=0.9.0, <0.10.0` cab be used. Sure, I can put together a PR.

**Comment by jkczyz** [comment]
Opened #77.

**Pull request merged**
